crates/sui-indexer-alt/src/tempdb.rs

```diff
@@ -0,0 +1,345 @@
+            }
+
+            std::thread::sleep(Duration::from_millis(50));
+        }
+
+        Err(HealthCheckError::Unknown(
+            "timeout reached when waiting for service to be ready".to_owned(),
+        ))
+    }
+}
+
+#[derive(Debug)]
+enum HealthCheckError {
+    NotRunning,
+    NotReady,
+    #[allow(unused)]
+    Unknown(String),
+}
+
+/// Run the postgres `pg_isready` command to get the status of database
+///
+/// See <https://www.postgresql.org/docs/16/app-pg-isready.html> for more info
+fn pg_isready(port: u16) -> Result<(), HealthCheckError> {
+    let output = Command::new("pg_isready")
+        .arg("--host=localhost")
+        .arg("-p")
+        .arg(port.to_string())
+        .arg("--username=postgres")
+        .output()
+        .map_err(|e| HealthCheckError::Unknown(format!("command not found: pg_ctl: {e}")))?;
+
+    trace!("pg_isready code: {:?}", output.status.code());
+    trace!("pg_isready output: {}", output.stderr.escape_ascii());
+    trace!("pg_isready output: {}", output.stdout.escape_ascii());
+    if output.status.success() {
+        Ok(())
+    } else {
+        Err(HealthCheckError::NotReady)
+    }
+}
+
+/// Run the postgres `initdb` command to initialize a database at the provided path
+///
+/// See <https://www.postgresql.org/docs/16/app-initdb.html> for more info
+fn initdb(dir: &Path) -> Result<()> {
+    let output = Command::new("initdb")
+        .arg("-D")
+        .arg(dir)
+        .arg("--no-instructions")
+        .arg("--username=postgres")
+        .output()
+        .context("command not found: initdb")?;
+
+    if output.status.success() {
+        Ok(())
+    } else {
+        Err(anyhow!(
+            "unable to initialize database: {:?}",
+            String::from_utf8(output.stderr)
+        ))
+    }
+}
+
+/// Return an ephemeral, available port. On unix systems, the port returned will be in the
+/// TIME_WAIT state ensuring that the OS won't hand out this port for some grace period.
+/// Callers should be able to bind to this port given they use SO_REUSEADDR.
+pub fn get_available_port() -> u16 {
+    const MAX_PORT_RETRIES: u32 = 1000;
+
+    for _ in 0..MAX_PORT_RETRIES {
+        if let Ok(port) = get_ephemeral_port() {
+            return port;
+        }
+    }
+
+    panic!("Error: could not find an available port");
+}
+
+fn get_ephemeral_port() -> std::io::Result<u16> {
+    // Request a random available port from the OS
+    let listener = std::net::TcpListener::bind(("127.0.0.1", 0))?;
+    let addr = listener.local_addr()?;
+
+    // Create and accept a connection (which we'll promptly drop) in order to force the port
+    // into the TIME_WAIT state, ensuring that the port will be reserved from some limited
+    // amount of time (roughly 60s on some Linux systems)
+    let _sender = std::net::TcpStream::connect(addr)?;
+    let _incoming = listener.accept()?;
+
+    Ok(addr.port())
+}
+
+#[cfg(test)]
```

Comment:
all the implementations above are copied from https://github.com/MystenLabs/sui/blob/main/crates/sui-indexer/src/tempdb.rs
made some minor changes below